src/coreclr/vm/methodtable.cpp

```diff
@@ -4026,6 +4026,391 @@ int MethodTable::GetRiscV64PassStructInRegisterFlags(CORINFO_CLASS_HANDLE cls)
+    CQuickArrayList<SwiftLoweringInterval> intervals;
+    for (uint32_t i = 0; i < loweredBytes.Size(); ++i)
+    {
+        // Don't create an interval for empty bytes
+        if (loweredBytes[i] == SwiftPhysicalLoweringTag::Empty)
+        {
+            continue;
+        }
+
+        bool startNewInterval =
+            // We're at the start of the type
+            i == 0
+            // We're starting a new float (as we're aligned)
+            || (IS_ALIGNED(i, 4) && loweredBytes[i] == SwiftPhysicalLoweringTag::Float)
+            // We're starting a new double or int64_t (as we're aligned)
+            || (IS_ALIGNED(i, 8) && (loweredBytes[i] == SwiftPhysicalLoweringTag::Double || loweredBytes[i] == SwiftPhysicalLoweringTag::Int64))
+            // We've changed interval types
+            || loweredBytes[i] != loweredBytes[i - 1];
+
+        if (startNewInterval)
+        {
+            SwiftLoweringInterval interval;
+            interval.m_offset = i;
+            interval.m_size = 1;
+            interval.m_tag = loweredBytes[i];
+            intervals.Push(interval);
+        }
+        else
+        {
+            intervals[intervals.Size() - 1].m_size++;
+        }
+    }
+
+    // Merge opaque intervals that are in the same pointer-sized block.
+    CQuickArrayList<SwiftLoweringInterval> mergedIntervals;
+
+    for (uint32_t i = 0; i < intervals.Size(); ++i)
+    {
+        SwiftLoweringInterval& interval = intervals[i];
+
+        if (interval.m_tag == SwiftPhysicalLoweringTag::Opaque)
+        {
+            // If we're at the start of the intervals, or the previous interval is not opaque, we need to start a new interval.
+            if (i == 0 || intervals[i - 1].m_tag != SwiftPhysicalLoweringTag::Opaque)
+            {
+                mergedIntervals.Push(interval);
+            }
+            // Otherwise, if the previous interval ends in the same pointer-sized block, we'll merge this interval into the previous one.
+            else if ((intervals[i - 1].m_offset + intervals[i - 1].m_size) / TARGET_POINTER_SIZE == interval.m_offset / TARGET_POINTER_SIZE)
+            {
+                SwiftLoweringInterval& lastInterval = mergedIntervals[mergedIntervals.Size() - 1];
+                lastInterval.m_size = interval.m_offset + interval.m_size - lastInterval.m_offset;
+            }
+        }
+        else
+        {
+            // Non-opaque intervals are never merged at this point.
+            mergedIntervals.Push(interval);
+        }
+    }
+
+    // Now we have the intervals, we can calculate the lowering.
+    CorInfoType loweredTypes[4];
+    uint32_t offsets[4];
+    uint32_t numLoweredTypes = 0;
+
+    for (uint32_t i = 0; i < mergedIntervals.Size(); i++, numLoweredTypes++)
+    {
+        SwiftLoweringInterval& interval = mergedIntervals[i];
+
+        if (numLoweredTypes == 4)
+        {
+            // If we have more than for intervals, this type is passed by-reference in Swift.
```

Comment:
```suggestion
            // If we have more than four intervals, this type is passed by-reference in Swift.
```

Please also make a doc reference here.